tests/cli.rs

```diff
@@ -0,0 +1,295 @@
+            assert_eq!(expect_content, get_output.stdout);
+            Ok(())
+        }
+        Some(out) => compare_files(path, out),
+    }
+}
+
+// Wrapping the [`Child`] process here allows us to impl the `Drop` trait ensuring the provide
+// process is killed when it goes out of scope.
+struct ProvideProcess {
+    child: Child,
+}
+
+impl Drop for ProvideProcess {
+    fn drop(&mut self) {
+        self.child.kill().ok();
+        self.child.try_wait().ok();
+    }
+}
+
+fn compare_files(expect_path: impl AsRef<Path>, got_dir_path: impl AsRef<Path>) -> Result<()> {
+    let expect_path = expect_path.as_ref();
+    let got_dir_path = got_dir_path.as_ref();
+    if expect_path.is_dir() {
+        let paths = std::fs::read_dir(expect_path)?;
+        for entry in paths {
+            let entry = entry?;
+            compare_files(entry.path(), got_dir_path)?;
+        }
+    } else {
+        let file_name = expect_path.file_name().unwrap();
+        let expect = std::fs::read(expect_path)?;
+        let got = std::fs::read(got_dir_path.join(file_name))?;
+        assert_eq!(expect, got);
+    }
+
+    Ok(())
+}
+
+/// Looks for regex matches on stderr output for the getter.
+///
+/// Errors on the first regex mis-match or if the stderr output has fewer lines than expected
+fn match_get_stderr(stderr: Vec<u8>) -> Result<()> {
+    let stderr = std::io::BufReader::new(&stderr[..]);
+    assert_matches_line![
+        stderr,
+        r"Fetching: [\da-z]{59}"; 1,
+        r"\[1/3\] Connecting ..."; 1,
+        r"\[2/3\] Requesting ..."; 1,
+        r"\[3/3\] Downloading collection..."; 1,
+        r"\d* file\(s\) with total transfer size [\d.]* ?[KMGT]?i?B"; 1,
+        r"Done in \d* seconds?"; 1
+    ];
+    Ok(())
+}
+
+/// Looks for regex matches on stderr output for the provider, returning the "all in one ticket"
+/// that can be used to 'get' from another process.
+///
+/// Errors on the first regex mismatch or if the stderr output has fewer lines than expected
+fn match_provide_stderr(
+    stderr: BufReader<ChildStderr>,
+    num_blobs: usize,
+    input: Input,
+) -> Result<String> {
+    // if we are using `stdin` we don't "read" any files, so the provider does not output any lines
+    // about "Reading"
+    let reading_line_num = if input == Input::Stdin { 0 } else { 1 };
+
+    let mut remaining_lines = assert_matches_line![
+        stderr,
+        r"Reading \S*"; reading_line_num,
+        r"Collection: [\da-z]{59}"; 1,
+        r""; 1,
+        r"- \S*: \d* bytes"; num_blobs,
+        r""; 1,
+        r"PeerID: [_\w\d-]*"; 1,
+        r"Auth token: [\w\d]*"; 1
+    ];
+
+    let re = r"All-in-one ticket: ([_a-zA-Z\d-]*)";
+    let rx = regex::Regex::new(re)?;
+    let line = next_line(&mut remaining_lines)?;
+    if !rx.is_match(&line) {
+        anyhow::bail!(match_err_msg(&line, re))
+    }
+    let caps = rx
+        .captures(&line)
+        .context("expected match on 'All-in-one' ticket")?;
+    Ok(caps
```

Comment:
if you make the macro return all captures from the regexps, you could add these to the macro invocation, which would read much nicer.